File: 0x10 基本数据结构/0x15 字符串/AcWing141.py
a, nxt = [0]*1000010, [0]*1000010
n, T = 0, 0

def calc_next():
    global n, T
    nxt[1], j = 0, 0
    for i in range(2, n+1):
        while j>0 and a[i]!=a[j+1]: j = nxt[j]
        if a[i] == a[j+1]: j += 1
        nxt[i] = j

def main():
    global n, T
    while True:
        n = int(input())
        if n == 0: break
        a[1:] = list(input())
        calc_next()
        T += 1; print("Test case #%d" %T)
        for i in range(2, n+1):
            if i%(i-nxt[i]) == 0 and i//(i-nxt[i]) > 1:
                print("%d %d" %(i, i//(i-nxt[i])))
        print()
        
main()

File: 0x10 基本数据结构/0x15 字符串/AcWing141.cpp
#include <iostream>
#include <cstring>
using namespace std;
char a[1000010];
int nxt[1000010], n, T;

void calc_next() {
  nxt[1] = 0;
  for (int i=2, j=0; i<=n; i++) {
    while (j>0 && a[i]!=a[j+1]) j = nxt[j];
    if (a[i] == a[j+1]) j++;
    nxt[i] = j;
  }
}

int main() {
  while (cin >> n && n) {
    scanf("%s", a+1);
    calc_next();
    printf("Test case #%d\n", ++T);
    for (int i=2; i<=n; i++)
      if (i%(i-nxt[i]) == 0 && i/(i-nxt[i]) > 1) // 能整除才是循环节
        printf("%d %d\n", i, i/(i-nxt[i]));
    puts("");
  }
}